[TDF] Registration of callbacks called on partial analysis results every N entries

bluehood opened the pull request:
Users can now register one or more callbacks to TResultProxies (i.e.
the results of TDF actions). A callback is just a callable that takes
a reference to the result type as argument and is going to be invoked
by each worker thread once every N entries (users choose N).

It is meant to be used to inspect partial results of the analysis
while the event loop is still running.
For example, in a single-thread event loop, one can draw a histogram
and update the canvas every 100 entries like this:
    
```c++
auto h = tdf.Histo1D("x");
TCanvas c("c","x hist");
// update the canvas every 100 entries
h.RegisterCallback(100, [&c](TH1D &h_) { h_.Draw(); c.Update(); });
// trigger event loop, this `Draw` will be performed afterwards
h->Draw();
```

Each worker thread invokes callbacks sequentially, but the same callback
might be invoked concurrently by different worker threads if implicit multi-threading
is enabled.

bluehood commented:
N.B.
Currently the mechanism is only "active" for non-jitted `Histo[1,2,3]D` actions.
For now TDF either throws an exception or prints a warning if users register a callback with the results of other actions.

pcanal commented:
>  but the same callback
might be invoked concurrently by different worker threads if implicit multi-threading
is enabled.

As far as I can tell this requires the user code to be thread safe, which does not seem to be the case of the example.

Also, in the blur, in it is not clear whether the call back will see the global partial result or the partial result so far on the 'current' thread.  The case 'global partial result' requires merging of the thread's histograms in the case there is one histo copy per thread/worker.

bluehood commented:
> As far as I can tell this requires the user code to be thread safe, which does not seem to be the case of the example.

Yes the example is meant to be single threaded. If users run the analysis on multiple threads they have to make sure the callback can be called by multiple threads concurrently.

> it is not clear whether the call back will see the global partial result or the partial result so far on the 'current' thread

No guarantees there: it's going to be a "partial result".
If it's "thread-local partial" or "global partial" depends on what each action's `PartialUpdate` method will do in practice. For some kind of actions it might be fine in terms of RAM and runtime to expose the "global" partial result, for others it could be an issue.

bluehood commented:
@pcanal here are full working examples with and without parallelism enabled:

**no IMT**
```c++
#include <thread>

#include "ROOT/TDataFrame.hxx"
#include "TApplication.h"
#include "TCanvas.h"
#include "TRandom.h"
using namespace ROOT::Experimental;

int main()
{
   TApplication app("app", nullptr, nullptr);

   // setup a TDF with some content
   TDataFrame d(5000);
   TRandom r;
   auto heavyWork = [&r]() {
      for (volatile int i = 0; i < 1000000; ++i) {}
      return r.Gaus();
   };
   auto tdf = d.Define("x", heavyWork);

   // get a histogram and register a callback
   auto h = tdf.Histo1D<double>({"", "", 100, -2., 2.}, "x");
   TCanvas c("c", "hist");
   h.RegisterCallback(100, [&c](TH1D &hist, unsigned int slot) {
      hist.Draw();
      c.Update();
   });

   // run analysis
   h->Draw();

   app.Run();
   return 0;
}
```

**IMT**
```c++
#include <algorithm>
#include <thread>
#include <chrono>

#include "ROOT/TDataFrame.hxx"
#include "TApplication.h"
#include "TCanvas.h"
#include "TRandom.h"
using namespace ROOT::Experimental;

int main()
{
   TApplication app("app", nullptr, nullptr);
   ROOT::EnableImplicitMT();
   const auto nSlots = ROOT::GetImplicitMTPoolSize();

   // setup a TDF with some content
   TDataFrame d(5000);
   std::vector<TRandom> rs(nSlots);
   auto heavyWork = [&rs](unsigned int slot) {
      for (volatile int i = 0; i < 1000000; ++i) {}
      return rs[slot].Gaus();
   };
   auto tdf = d.DefineSlot("x", heavyWork);

   // get a histogram and register a callback
   auto h = tdf.Histo1D<double>({"", "", 100, -2., 2.}, "x");
   TCanvas c("c", "running...");
   const auto mainThreadID = std::this_thread::get_id();
   h.RegisterCallback(100, [&c, &mainThreadID](decltype(h)::Value_t &hist, unsigned int slot) {
      if (std::this_thread::get_id() == mainThreadID) {
         hist.Draw();
         c.Update();
      }
   });
   // run analysis in parallel
   h->Draw();

   app.Run();
   return 0;
}
```

bluehood commented:
@Axel-Naumann @pcanal @dpiparo 
What do you think about changing the behaviour so that only one thread calls the callbacks?
So users don't have to worry about concurrency which might be a useless complication.
For histograms we wouldn't want to do synchronized partial merges of the thread-local copies anyway, I guess...

pcanal commented:
> No guarantees there: it's going to be a "partial result".
> If it's "thread-local partial" or "global partial" depends on what each action's

I am afraid that this is going to be confusing/hard-to-explain ....

Axel-Naumann commented:
> only one thread calls the callbacks

That's actually the behavior I had expected. I'd even argue that the main thread should register its histograms with ROOT, e.g. in a sub-`TDirectory` of gROOT, e.g. through a function call (`TDF::Publish("NameOfgROOTSubDir")`). These histograms should then be the merge target.

bluehood commented:
> it's going to be a "partial result".
> If it's "thread-local partial" or "global partial" depends on what each action's

> I am afraid that this is going to be confusing/hard-to-explain .... On the other hand in some debugging scenario it might be helpful

I would just leave the explanation at: "you get a partial analysis result (histogram filled with a part of the selected events, counter incremented only up to a certain point, mean over a subset of the events and so forth)". I think this is enough to enjoy the feature and use it correctly.

> only one thread calls the callbacks

> That's actually the behavior I had expected. 

Fair enough. If no one has anything against it, I will change the behavior to have only one thread call the callback.

@Axel-Naumann the event loop is still blocking so I'm not sure what users could do with the histograms we register with ROOT?

Axel-Naumann commented:
Throw in a couple of `gSystem->ProcessEvents()`, maybe? :-) No idea whether that's enough to make it usable.

dpiparo commented:
As a side note, the feature of "snapshot merging"/"peeping at results in the making" is something we had already in the first implementation of TThreadedObject (see https://root.cern/doc/master/mt201__parallelHistoFill_8C.html). This was a rudimental building block, which needed the user to asynchronously invoke a method. I see this functionality as something quite interesting. In general, "callbacks"/"hooks" of different sorts are a fundamental piece of data processing frameworks of which TDF is de facto an example. I welcome this feature warmly, also when thinking about the direction we have taken with the datasource, which consists in giving users the possibility to customise the usage of TDF according to their needs - we provide a rocksolid platform to solve their problems in an optimised way.

A Crazy (?) example: right now we would be able to:
- Write a datasource that reads measurements coming from an oscilloscope/simple detector
- TDF loops and processes these measurements
- Histograms are produced, among other kind of data, and shown while refreshing in real time
- A file with the data acquired and manipulated is produced

dpiparo commented:
As far as I am concerned, this is ready to go.

pcanal commented on tree/treeplayer/inc/ROOT/TDFNodes.hxx:
```diff
@@ -105,6 +130,7 @@ class TLoopManager : public std::enable_shared_from_this<TLoopManager> {
    std::string fToJit;        ///< string containing all `BuildAndBook` actions that should be jitted before running
    const std::unique_ptr<TDataSource> fDataSource; ///< Owning pointer to a data-source object. Null if no data-source
    ColumnNames_t fDefinedDataSourceColumns; ///< List of data-source columns that have been `Define`d so far
+   std::vector<TLoopCallback> fCallbacks; ///< Registered callbacks invoked once per event per slot
```
If it is invoke once per event per slot (so once per event), what is the semantic difference with an (unfiltered) part of the functional chain applied to each events?

pcanal replied:
Humm ... Actually maybe the comment is obsolete as it seems to not match well with the "Never call callbacks concurrently, only ever run on slot 0" ....

bluehood replied:
Yes the comment there is wrong. I made several changes to the behavior of the feature after your comments as well as other people's. New revised version is coming in a minute :)

bluehood commented:
Thanks everyone for the valuable feedback! I changed the behaviour of the feature to make it simpler to use correctly, and also added one or two perks:

* `OnParameterUpdate` (was `RegisterCallback`) now takes a callable that will be invoked once every specified number of entries on a partial result in _one_ of the worker threads. It will never be invoked concurrently
* `OnParameterUpdateSlot` takes a callable that will be invoked once every specified number of entries on a partial result in each of the worker threads. It will be invoked concurrently, and will also take as argument the "processing slot" number the partial result belongs to
* passing `0` as `everyNEvents` parameter to `OnParameterUpdate[Slot]` makes it so that the callback is invoked just once [per slot] before starting the event loop

bluehood commented:
@pcanal now users can register a drawing of a result histogram in the same way in single- and multi-thread analysis, like this:

```c++
   // Draw partial result on canvas every 500 events
   TCanvas c("c", "Running event loop...");
   h.OnPartialResult(100, [&c](TH1D &h_) {
      c.cd();
      h_.Draw();
      c.Update();
   });
```
@Axel-Naumann users can open a `TBrowser` and check result updates while the event loop running quite easily:
```c++
// create "TDFResults" directory in TBrowser
   TBrowser b("b", "event loop peeper");
   TMemFile m("TDFResults", "RECREATE");

   // add partial result to the "TDFResults"
   m.Browse(&b); // it would be cool if we could change the current directory in the TBrowser to `TDFResults` here
   h.OnPartialResult(decltype(h)::kOnce, [&m](TH1D &h_) { m.Add(&h_); });

   // call ProcessEvents every once in a while during the event loop to allow users to navigate the TBrowser
   h.OnPartialResult(50, [](TH1D &hist) {
      gSystem->ProcessEvents();
   });
```
@peremato here is how one would implement a thread-safe progress bar for a TDF multi-thread analysis
```c++
   // Update progress bar every 100 events
   std::string progress;
   std::mutex bar_mutex;
   c.OnPartialResultSlot(nEvents / 100, [&progress, &bar_mutex](unsigned int, ULong64_t &c_) {
      std::lock_guard<std::mutex> lg(bar_mutex);
      progress.push_back('#');
      std::cout << "\r[" << std::left << std::setw(100) << progress << ']' << std::flush;
   });
```

Here are the corresponding self-contained, fully working code examples:
[draw_partial_result.txt](https://github.com/root-project/root/files/1327464/draw_partial_result.txt)
[inspect_analysis.txt](https://github.com/root-project/root/files/1327466/inspect_analysis.txt)
[progress_bar.txt](https://github.com/root-project/root/files/1327469/progress_bar.txt)
(for some reason github does not allow files with `.cpp` extension :man_shrugging: )

For common-enough callbacks one can always imagine to provide helper functions of course.
I like this feature a lot. What do you think?

bluehood commented:
The failure in `test_misc` is fixed by PR #1046 and a similar PR to roottest

bluehood commented:
Rebased on master now that #1046 has been merged. Tests are expected to be green now (although more testing of the callback registration has to be added).

pcanal commented on tree/treeplayer/test/dataframe/dataframe_callbacks.cxx:
```diff
@@ -0,0 +1,72 @@
+   using value_t = typename decltype(h)::Value_t;
+   ULong64_t i = 0ull;
+   ULong64_t everyN = 1ull;
+   h.OnPartialResult(everyN, [&i, &everyN](value_t &h_) {
```
Should this be extended to also test OnPartialResultSlot. ?

bluehood replied:
yes, and all the different actions! I will lift the `do not merge` once support for jitted actions and all the necessary tests are in.

dpiparo commented:
By the way, the callbacks can also become a way in which profiling can be performed. For example, we can measure time, memory, cache misses or dump something more complete (e.g. an igprof report) *during* the event loop. This kind of instrumentation will be very useful.

bluehood commented:
Rebase on master to fix conflicts.
I will definitely add more multi-thread tests but I think this can be safely merged if jenkins returns green.

bluehood commented:
I'll have to check what is causing these failures on mac, re-adding "do not merge" label :sweat:

bluehood commented:
@phsft-bot build

bluehood commented:
Rebase on master + attempt at making mac1012 happy

bluehood commented:
Uhm....so this gtest:
```c++
auto m = tdf.Min<double>("x");
double runningMin = std::numeric_limits<double>::max();
m.OnPartialResult(2, [&runningMin](double x) {
  runningMin = x;
});
EXPECT_DOUBLE_EQ(runningMin, *m); 
```
results in this error:
```
[ RUN      ] TDFCallbacks.Min
/Volumes/MacintoshHD/build/jenkins/workspace/root-pullrequests-build/root/tree/treeplayer/test/dataframe/dataframe_callbacks.cxx:126: Failure
      Expected: runningMin
      Which is: 1.7976931348623157e+308
To be equal to: *m
      Which is: -3.0919523479251314
[  FAILED  ] TDFCallbacks.Min (0 ms)
```
only on mac1012 and only for `Min` and `Max` which have the same kind of callback logic.
Do lambda captures work differently on mac? I'm stumped.

bluehood commented:
Ah, I fell for the oldest trick in the book!
Only on mac, in `EXPECT_DOUBLE_EQ(runningMin, *m)` the first argument was evaluated before the second, but the evaluation of the second argument filled the first with the correct result.
